Add MapGuard to automatically unmap virtual memory

dubek opened the pull request:
During work on the attestation protocol I noticed that writing functions that map several memory areas into virtual memory is very cumbersome due to the need to unmap these areas in every error/exit condition throughout the function.

I introduced a MapGuard struct (similar to LockGuard) -- it implements the `Drop` trait which unmaps the area when the MapGuard goes out of scope. This allows mapping several areas with MapGuard and then returning whenever is needed (using `return` or `?`) from the middle of the function without manually unmapping all the still-mapped areas.

Example:

```rust
fn work() -> Result<()> {
  let map1 = MapGuard::new_private(gpa, size, true)?;
  // view the memory as a C struct
  let req: &MyRequestStruct = map1.as_object();

  let map2 = MapGuard::new_private(gpa2, size, true)?; // <--- an error here will cause map1 to unmap
  // view the memory as a slice of bytes with the correct size (the entire mapped area)
  let buf: &[u8] = map2.as_bytes();

  // read from the mapped memory
  if some_condition {
    return Err(...); // here both areas are unmapped
  }

  Ok(())
  // here both areas are unmapped
}
```

The second commit replaces one pair of `pgtable_map_pages_private()/pgtable_unmap_pages()` with MapGuard. There are other opportunities to use it in linux-svsm, but:
1. I wanted to get feedback first
2. Changes in svsm_request will conflict with #38 , so I prefer to wait.

Zildj1an commented on src/cpu/smp.rs:
```diff
@@ -243,14 +243,12 @@ pub fn smp_run_bios_vmpl() -> bool {
         }
 
-        let vmsa_va: VirtAddr = match pgtable_map_pages_private(vmsa_pa, PAGE_SIZE) {
+        let vmsa_map: MapGuard = match MapGuard::new_private(vmsa_pa, PAGE_SIZE, true) {
```
Because almost all invocations to this function will have `unmap_on_drop=true`, let's assume that's the case and have other function `new_private_persistent()` that sets it to false.

Zildj1an commented on src/cpu/smp.rs:
```diff
@@ -243,14 +243,12 @@ pub fn smp_run_bios_vmpl() -> bool {
-        vc_ap_create(vmsa_va, PERCPU.apic_id());
-
-        pgtable_unmap_pages(vmsa_va, PAGE_SIZE);
+        vc_ap_create(vmsa_map.va, PERCPU.apic_id());
```
Use `funcs!` in the implementation of MapGuard to retrieve the VA using the getter (`vmsa_map.va()`). Fields `va` and `len` will not need to be pub anymore.

dubek replied:
You're right, I'll change that.

BTW - for another issue - [Rust naming conventions](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter) say that getters should not begin with `get_`. Not sure if we want to adhere to that.

Zildj1an commented on src/mem/map_guard.rs:
```diff
@@ -0,0 +1,75 @@
+use x86_64::{PhysAddr, VirtAddr};
+
+/// An area mapped into virtual memory. If `unmap_on_drop` is true, the
+/// area is unmapped when the `MapGuard` is dropped (out of scope).
```
Add the example of usage you include in the PR here, as part of the comments

Zildj1an reviewed (changes requested):
Thanks for this @dubek very cool idea. I added some comments before we can merge. Besides those, also:

1. We need to have all the `pgtable_map_pages_xxx()` replaced prior to merge.
2. Be cognizant that in the case of `fwcfg_map_bios()` we need to return the MapGuard object and the caller will use it. 
3. Let's remove the pub from `pgtable_map_pages_xxx()` to make sure everyone uses this approach from now on. 
 
Feel free to add my Reviewed-by tag to your updated commits!

dubek commented:
Thanks @Zildj1an for the review. I modified the first two commits to address the comments.
I'll now start adding additional commits to replace `pgtable_map_pages_xxx()` invocations.

dubek commented:
I rebased and made the requested changes.

In some places more invasive refactoring was needed, so please review again. I tried to separate the changes by commits.

For example - in commit b4eca3a9f4dc3 - in `fwcfg_map_bios()`, instead of returning a MapGuard, I renamed the function to `fwcfg_get_bios_area()`, and it no longer maps the area. I made another function `prepare_bios()` which maps the bios, prepare the needed permissions, and unmaps it (and that function uses MapGuard).

tlendacky reviewed (commented):
It would be easier to review if you split this into two patches. The first patch creates the helper function and the second patch makes the MagGuard change.

tlendacky commented on src/protocols/core.rs:
```diff
@@ -415,48 +421,48 @@ unsafe fn handle_pvalidate_request(vmsa: *mut Vmsa) {
-    let mut e_va: VirtAddr = va + size_of::<PvalidateRequest>();
-    while (*request).next() < (*request).entries() {
-        let entry: *const PvalidateEntry = e_va.as_ptr();
+    {
```
Maybe I'm missing something, why do the braces need to be added?  I guess I don't understand why they're there and it makes the diff much harder to read.

dubek replied:
Maybe try to view the diff while ignoring whitespace (`git diff -w`, and there's also an option in the github UI) can help with the indent change.

It's because we need to unmap the page before the end of the function where we do:

```
    if flush {
        invlpgb_all();
        tlbsync();
    }
```

But now I think -- maybe I'll just call `drop(map);` ([doc](https://doc.rust-lang.org/std/mem/fn.drop.html)) in the original location of the unmap, instead of insisting on the indented scopes with braces.

tlendacky replied:
> But now I think -- maybe I'll just call `drop(map);` ([doc](https://doc.rust-lang.org/std/mem/fn.drop.html)) in the original location of the unmap, instead of insisting on the indented scopes with braces.

It doesn't look like you did this.

dubek replied:
Yes, I forgot about the pvalidate patch. I'll change it as well.

dubek commented on src/bios.rs:
```diff
@@ -254,31 +254,30 @@ unsafe fn advertise_svsm_presence(bios_info: &mut BiosInfo, caa: PhysAddr) -> bo
-    (*bios_secrets).set_svsm_guest_vmpl(1);
-
-    pgtable_unmap_pages(bios_secrets_va, section.size_u64());
+    {
```
Here it's probably OK to keep both pages mapped until the end of the function (or any early return). I'll change it.

In other places, this might not be the case. For example in `handle_create_vcpu_request()` we need to unmap the vmsa page and then call `invlpgb_all()` and `tlbsync()`. If we don't put the explicit brackets, the drop of the MapGuard will occur at the end of the function *after* `invlpgb_all()` and `tlbsync()`, modifying the behaviour. If we want to avoid the extra indent we can extract the code that deals with the mapped page to a helper function, but that might be cumbersome if there are a lot of stack argument we need to pass to that function.

dubek commented:
> It would be easier to review if you split this into two patches. The first patch creates the helper function and the second patch makes the MagGuard change.

I see this comment in the top-level conversation only; which patch / function is it referring to?

tlendacky commented:
> > It would be easier to review if you split this into two patches. The first patch creates the helper function and the second patch makes the MagGuard change.
> 
> I see this comment in the top-level conversation only; which patch / function is it referring to?

Sorry, that was referring to the handle_create_vcpu_request() patch.

tlendacky commented on src/protocols/core.rs:
```diff
@@ -268,80 +327,31 @@ unsafe fn handle_create_vcpu_request(vmsa: *mut Vmsa) {
+        Err(code) => {
+            // On error turn the page (back) into a non-VMSA page
+            grant_vmpl_access(create_vmsa_va, RMP_4K, vmpl as u8);
+            code
```
And this still needs to go through the tlb flush instead of returning directly.

dubek replied:
It should go all the way to the end; this `match` statement just sets the `ret` variable, which is then used at the end `(*vmsa).set_rax(ret);`.

tlendacky replied:
Ah, my bad...  confused it with the implicit return thing. One of the things about Rust that I'm not crazy about, you don't always have to use "return" and so I misread this and forgot it is in a match statement.

tlendacky commented on src/protocols/core.rs:
```diff
@@ -247,6 +247,65 @@ unsafe fn handle_delete_vcpu_request(vmsa: *mut Vmsa) {
     (*vmsa).set_rax(SVSM_SUCCESS);
 }
 
+unsafe fn create_vcpu_inner(
```
I'd still like to see this named ```__handle_vcpu_create_request()```

tlendacky commented on src/mem/map_guard.rs:
```diff
@@ -0,0 +1,113 @@
+}
+
+impl Drop for MapGuard {
+    fn drop(&mut self) {
```
Quick question, if drop() is called directly, what happens when the map goes out of scope, won't this be called a second time and result in possible error messages from pgtable_unmap_pages()?

Just wondering if there needs to also be a boolean that indicates if the pages are still mapped.

dubek replied:
> Quick question, if drop() is called directly, what happens when the map goes out of scope, won't this be called a second time and result in possible error messages from pgtable_unmap_pages()?
> 
> Just wondering if there needs to also be a boolean that indicates if the pages are still mapped.

The way I understand it, the manual `drop(x)` _is_ the end of the scope for that variable (official, it is moved into the drop function); so memory will be unmapped at that point and not again. Compiler should also complains if you try to use `map` after it was dropped, which is nice. See the example code: https://doc.rust-lang.org/rust-by-example/trait/drop.html (you can edit and run it in the browser; press the ▶️ button at the top-right corner of the code).

tlendacky replied:
Excellent! Thanks for educating me on this!
